lib/common/bits.h

```diff
@@ -0,0 +1,244 @@
+}
+
+MEM_STATIC unsigned ZSTD_countTrailingZeros64(U64 val)
+{
+    assert(val != 0);
+#   if defined(_MSC_VER) && defined(_WIN64)
+#       if STATIC_BMI2
+            return _tzcnt_u64(val);
+#       else
+            if (val != 0) {
+                unsigned long r;
+                _BitScanForward64(&r, val);
+                return (unsigned)r;
+            } else {
+                /* Should not reach this code path */
+                __assume(0);
+            }
+#       endif
+#   elif defined(__GNUC__) && (__GNUC__ >= 4)
+        if (MEM_32bits()) {
+            U32 mostSignificantWord = (U32)(val >> 32);
+            U32 leastSignificantWord = (U32)val;
+            if (leastSignificantWord == 0) {
+                return 32 + (unsigned)__builtin_ctz(mostSignificantWord);
+            } else {
+                return (unsigned)__builtin_ctz(leastSignificantWord);
+            }
+        } else {
+            return (unsigned)__builtin_ctzll(val);
+        }
+#   else
+        return ZSTD_countTrailingZeros64_fallback(val);
+#   endif
+}
+
+MEM_STATIC unsigned ZSTD_NbCommonBytes_fallback(size_t val)
+{
+    if (MEM_isLittleEndian()) {
+        if (MEM_64bits()) {
+            static const int DeBruijnBytePos[64] = { 0, 0, 0, 0, 0, 1, 1, 2,
+                                                     0, 3, 1, 3, 1, 4, 2, 7,
+                                                     0, 2, 3, 6, 1, 5, 3, 5,
+                                                     1, 3, 4, 4, 2, 5, 6, 7,
+                                                     7, 0, 1, 2, 3, 3, 4, 6,
+                                                     2, 6, 5, 5, 3, 4, 5, 6,
+                                                     7, 1, 2, 4, 6, 4, 4, 5,
+                                                     7, 2, 6, 5, 7, 6, 7, 7 };
+            return DeBruijnBytePos[((U64)((val & -(long long)val) * 0x0218A392CDABBD3FULL)) >> 58];
+        } else { /* 32 bits */
+            static const int DeBruijnBytePos[32] = { 0, 0, 3, 0, 3, 1, 3, 0,
+                                                     3, 2, 2, 1, 3, 2, 0, 1,
+                                                     3, 3, 1, 2, 2, 2, 2, 0,
+                                                     3, 1, 2, 0, 1, 0, 1, 1 };
+            return DeBruijnBytePos[((U32)((val & -(S32)val) * 0x077CB531U)) >> 27];
+        }
+    } else {  /* Big Endian CPU */
+        unsigned r;
+        if (MEM_64bits()) {
+            const unsigned n32 = sizeof(size_t)*4;   /* calculate this way due to compiler complaining in 32-bits mode */
+            if (!(val>>n32)) { r=4; } else { r=0; val>>=n32; }
+            if (!(val>>16)) { r+=2; val>>=8; } else { val>>=24; }
+            r += (!val);
+            return r;
+        } else { /* 32 bits */
+            if (!(val>>16)) { r=2; val>>=8; } else { r=0; val>>=24; }
+            r += (!val);
+            return r;
+        }
+    }
+}
+
+MEM_STATIC unsigned ZSTD_NbCommonBytes(size_t val)
```

Comment:
nit: Add a `ZSTD_countLeadingZeros{32, 64}()` function.

Then you can implement `ZSTD_NbCommonBytes()` without any `#ifdef` code or fallbacks.

The only difference is that the software fallback would need one extra `>> 3` instruction. But I don't think that is a big deal.